Added the gravity compensation feature to the Aloha stack

Shiming-Liang opened the pull request:
Major changes:
- A client class in the `robot_utils.py` to enable/disable gravity compensation
- New entries in the launch file to bring up the gravity compensation nodes
- A `leader_motor_specs.yaml` for the gravity compensation nodes
- Slight changes to the `dual_side_teleop.py` and the `sleep.py` scripts to switch between torque on/off and enable/disable gravity compensation

Shiming-Liang committed:
added a service client to enable/disable gravity compensation

Shiming-Liang committed:
added switch between grav comp or torque off

Shiming-Liang committed:
use namespace instead of param for gravity compensation in the launch file

Shiming-Liang committed:
fixed typo in the grav comp config

Shiming-Liang committed:
updated the motor_specs configuration with more sophisticated friction compensation options

Shiming-Liang committed:
Minor fixed for the gravity compensation feature

launch
- changed the path concatenation of the motor specs for better readability

Shiming-Liang committed:
improved the gravity compensation feature

- unified the style of the enable/disable gravity compensation function with the others like torque_on/off
- added the feature for the record_episodes script

Shiming-Liang committed:
Minor fixes

- fixed flake8 test failures
- changed the behavior of the dual_teleop script to be similar to the record_episode script
- removed checking for the gravity compensation nodes at the sleep script, assuming it's running as per the default launch script

lukeschmitt-tr commented on config/leader_motor_specs.yaml:
Let's make two files - leader_left and leader_right. We have seen how two different robots of the same model can act slightly differently — we might want to tune each arm individually.

lukeschmitt-tr commented on scripts/dual_side_teleop.py:
```diff
@@ -12,8 +14,9 @@
     get_arm_gripper_positions,
     move_arms,
     move_grippers,
-    torque_off,
     torque_on,
+    torque_off,
+    enable_gravity_compensation,
```
alphabetize

lukeschmitt-tr commented on scripts/dual_side_teleop.py:
```diff
@@ -154,4 +164,10 @@ def main() -> None:
-    main()
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--gravity_compensation',
```
```suggestion
        '-g', '--gravity_compensation',
```

lukeschmitt-tr commented on scripts/dual_side_teleop.py:
```diff
@@ -154,4 +164,10 @@ def main() -> None:
+    parser.add_argument(
+        '--gravity_compensation',
+        action='store_true',
+        help='If set, gravity compensation will be enabled for the leader robots.',
```
If set, gravity compensation will be enabled for the leader robots when teleop starts.

lukeschmitt-tr commented on scripts/record_episodes.py:
```diff
@@ -28,6 +28,8 @@
     Recorder,
     torque_off,
     torque_on,
+    enable_gravity_compensation,
+    disable_gravity_compensation,
```
alphabetize

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -293,3 +296,34 @@ def postprocess_base_action(base_action):
     return np.array([linear_vel, angular_vel])
+
+
+class GravityCompensationClient(Node):
```
Refactor this to xs_modules as per our discussion offline

lukeschmitt-tr commented on scripts/record_episodes.py:
```diff
@@ -395,4 +410,9 @@ def debug():
         ),
     )
+    parser.add_argument(
+        '--gravity_compensation',
```
```suggestion
        '-g','--gravity_compensation',
```

lukeschmitt-tr commented on scripts/record_episodes.py:
```diff
@@ -395,4 +410,9 @@ def debug():
+    parser.add_argument(
+        '--gravity_compensation',
+        action='store_true',
+        help='If set, gravity compensation will be enabled for the leader robots.',
```
If set, gravity compensation will be enabled for the leader robots when teleop starts.

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -293,3 +296,34 @@ def postprocess_base_action(base_action):
+        rclpy.spin_until_future_complete(self, self.cli.call_async(self.req))
+
+
+def enable_gravity_compensation(bot: InterbotixManipulatorXS):
+    client = GravityCompensationClient(bot.core.robot_name)
+    client.enable()
+    client.destroy_node()
+
+
+def disable_gravity_compensation(bot: InterbotixManipulatorXS):
+    client = GravityCompensationClient(bot.core.robot_name)
+    client.disable()
+    client.destroy_node()
```
These simple calls can stay

Shiming-Liang committed:
Updates on the gravity compensation feature

- alphabetized the imports
- used two separate motor specs files for the leader left and right arm
- added a shorthand for the script argument for gravity compensation

Shiming-Liang committed:
Updated the auto_record.sh

- removed venv check since it's only used by ACT in training and evaluation
- added options for enabling base torque and gravity compensation

Shiming-Liang committed:
Moved the interface with gravity compensation to interbotix_xs_modules

lukeschmitt-tr commented on scripts/auto_record.sh:
```diff
@@ -1,17 +1,17 @@
 ROS_DISTRO=humble
 
-VENV_ACTIVATE_PATH=$HOME/act/bin/activate
+# VENV_ACTIVATE_PATH=$HOME/act/bin/activate
```
Remove commented lines if not necessary to be there

lukeschmitt-tr commented on scripts/auto_record.sh:
```diff
@@ -31,7 +31,7 @@ echo "Task: $1"
 do
   echo "Starting episode $i"
-  python3 "$RECORD_EPISODES" --task "$1"
+  python3 "$RECORD_EPISODES" --task "$1" $3 $4
```
Put args in double quotes

```suggestion
  python3 "$RECORD_EPISODES" --task "$1" "$3" "$4"
```

Shiming-Liang replied:
The python script will complaint about getting `python3 record_episodes.py --task "xxx" "-g"`. But we can remove the "" for $1 to keep the style consistent. What do you think?

lukeschmitt-tr replied:
Sure, that works for me.

Shiming-Liang replied:
I looked into this more. It was the optional argument that caused the problem. We are getting `python3 record_episodes.py --task "xxx" "-g" ""`, where the last one is a empty str not defined in the arg parser. Either way, not using the quotes seems to be a good solution.

Shiming-Liang committed:
Minor update to the auto_record.sh

- removed commented lines not used
- removed quotes for $1 to align the style
